[core] Properly recycle `Device::temp_suspected`.

jimblandy opened the pull request:
Change the various uses of `Device::temp_suspected` to correctly
recycle its allocations when possible.

In the absence of errors, `Device::temp_suspected` is only used as
temporary storage within particular function calls, and is a field of
`Device` only as a way to recycle its allocations.

However, the code in `Global::queue_submit` that tries to use
`Device::temp_suspected` in this way doesn't quite pull it off.
It unconditionally replaces any prior `ResourceMaps` stored there
with a fresh empty one, clears the prior one, and then drops it,
wasting its allocations.

Rather than discard everything in `temp_selected` on every call to
`Device::maintain`, change `ResourceMaps::extend` to take the source
via `&mut` rather than by value, and have callers only obtain a `&mut`
to `Device::temp_selected`, rather than using `Option::take` to take
ownership of the value.

ErichDonGubler commented:
I'll take on review for this tomorrow. 👍🏻

ErichDonGubler commented:
I promised review for yesterday, but it looks like this will be in my queue until Monday. Sorry for the wait! 😖

ErichDonGubler reviewed (changes requested):
This PR conflicts with <https://github.com/gfx-rs/wgpu/pull/5653>, it seems, particularly my thought about [removing `temp_suspected` altogether](https://github.com/gfx-rs/wgpu/pull/5653#discussion_r1589505878). I think this should be closed? 🤔

jimblandy commented:
Yes, most of this PR is obviated by #5653, but the change to `Device::untrack` still makes sense.

ErichDonGubler commented:
@jimblandy: Weeeell then let's rebase and re-review! 😀

ErichDonGubler reviewed (approved):
Approving contingent on the rebase that we've agreed on.

jimblandy committed:
[core] Properly recycle `Device::temp_suspected`.

Change `Device::untrack` to properly reuse the `ResourceMap` allocated
for prior calls. The prior code tries to do this but always leaves
`Device::temp_suspected` set to a new empty `ResourceMap`, leaving the
previous value to be dropped by `ResourceMap::extend`.

Change `ResourceMap::extend` to take `other` by reference, rather than
taking it by value and dropping it.
